### PR TITLE
feat: tasks to push data to table in postgres

### PR DESCRIPTION
**Pull request by tombrooks248**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adding tasks to task table in Postgres in addition to delta lake.

ensuring duplicate tasks are not created from logs.

Adding logging (hidden behind a debug flag) to keep track of how many rows are being added to delta lake and the Postgres table

Also adding logging to track how long it takes to figure out s3 paths because this took a while in PROD last night and I want to see if I am doing this particularly inefficiently.

refactored the write_postgres function to use a staging table and atomic change so no change of losing data if an error occurs mid write process.

Using Anypath to create the s3 paths to make the code neater.

Added an integration test to catch any code changes which might add duplicate tasks.

## Related Tickets & Documents

- [Ticket Link](https://github.com/digital-land/pyspark-jobs/issues/59)
- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

I have run this several times in staging it is creating the rows we need and producing logging so it should be easy to follow what is happening and diagnose any issues in future.

Example of the logging when the pipeline runs: 

image

Example of the logging to show the pipeline no longer produces duplicates: 

image

## Added/updated tests?
_We encourage you to keep the code coverage percentage at 80% and above. Please refer to the [Digital Land Testing Guidance](https://digital-land.github.io/technical-documentation/development/testing-guidance/) for more information._

- [x] Yes
- [] No, and this is why
- [ ] I need help with writing tests

### CODE REVIEW

File: src/jobs/pipeline.py [hunk 1]
```diff
@@ -748,12 +748,59 @@ def execute(self):
             else reduce(lambda a, b: a.unionByName(b), frames)
         )
 
+        count_before = tasks_df.count()
```
Comment (eveleighoj) [inline src/jobs/pipeline.py]: counts force computation in spark so generally might be worth only printing if ran with a debug flag of some sort.

File: src/jobs/pipeline.py [hunk 2]
```diff
@@ -748,12 +748,59 @@ def execute(self):
...
+        tasks_df = tasks_df.dropDuplicates(["reference"])
```
Comment (eveleighoj) [inline src/jobs/pipeline.py]: by how we're pulling in the data there should be no duplicate tasks instead of adding extra computation here I would add assertions to the tests or add additional tests which check that tasks can't be duplicates

Comment (eveleighoj) [reply src/jobs/pipeline.py]: reading through this you could add a debug flag for this job that can be used in dev airflow and an option on staging/prod and this block could be a debug thing

File: src/jobs/pipeline.py [hunk 3]
```diff
@@ -748,12 +748,59 @@ def execute(self):
...
         )
-        logger.info("TaskPipeline: Complete")
+        logger.info(f"TaskPipeline: Delta table written to {output_path}")
+
+        if self.config.database_url:
+            logger.info(f"TaskPipeline: Writing {count_after} tasks to Postgres...")
+            self._write_postgres(tasks_df)
+
+    def _write_postgres(self, tasks_df):
+        import pg8000
```
Comment (eveleighoj) [inline src/jobs/pipeline.py]: imports at top of file, Claude seems to hate doing this for some reason

File: src/jobs/pipeline.py [hunk 4]
```diff
@@ -748,12 +748,59 @@ def execute(self):
...
             .save(output_path)
         )
-        logger.info("TaskPipeline: Complete")
+        logger.info(f"TaskPipeline: Delta table written to {output_path}")
+
+        if self.config.database_url:
+            logger.info(f"TaskPipeline: Writing {count_after} tasks to Postgres...")
+            self._write_postgres(tasks_df)
+
+    def _write_postgres(self, tasks_df):
```
Comment (eveleighoj) [inline src/jobs/pipeline.py:721]: worth looking at postgres writer functions in the utils folder. Might be an opportunity to reduce duplicate code

File: src/jobs/pipeline.py [hunk 5]
```diff
@@ -748,12 +748,59 @@ def execute(self):
...
+        import pg8000
+        from pyspark.sql.types import DateType
+
+        conn_params = parse_database_url(self.config.database_url)
+
+        tasks_df = tasks_df.withColumn("entry_date", col("entry_date").cast(DateType()))
+
+        url = f"jdbc:postgresql://{conn_params['host']}:{conn_params['port']}/{conn_params['database']}"
+        tasks_df.write.jdbc(
```
Comment (eveleighoj) [inline src/jobs/pipeline.py]: writer functions general follow a pattern of creating and loading to a staging table and then swapping data in. Worth a bit of research but I think it's better for the db

File: src/jobs/pipeline.py [hunk 6]
```diff
@@ -748,12 +748,59 @@ def execute(self):
...
+                "driver": "org.postgresql.Driver",
+                "truncate": "true",
+                "stringtype": "unspecified",
+                "batchsize": "5000",
+                "reWriteBatchedInserts": "true",
+            },
+        )
+        logger.info("TaskPipeline: Wrote tasks to Postgres")
+
+        conn = pg8000.connect(**conn_params)
```
Comment (eveleighoj) [inline src/jobs/pipeline.py]: again might be better as a debug thing to reduce the ask on the database day to day

File: src/jobs/pipeline.py [hunk 7]
```diff
@@ -649,25 +650,31 @@ def execute(self):
...
                 p
                 for p in _list_s3_paths(bucket, s3_prefix, "collection/resource.csv")
                 if "-collection/collection/resource.csv" in p
             ]
+            logger.info(f"TaskPipeline: Found {len(resource_files)} resource files")
+
+            logger.info("TaskPipeline: Listing log files")
             log_files = [
                 p
                 for p in _list_s3_paths(bucket, s3_prefix, "collection/log.csv")
```
Comment (eveleighoj) [inline src/jobs/pipeline.py]: I know this is already in but worth checking what cloudpathlib can help simplify this code. it aims to simulate. pathlib which I think can do these types of things using globs

**Review by eveleighoj — CHANGES_REQUESTED** [review]

File: src/jobs/pipeline.py (hunk 1 above)
Comment (tombrooks248) [reply src/jobs/pipeline.py]: Put most of the counting behind a debug flag - this specific one is removed now I got to the bottom of the duplicates issue.

File: src/jobs/pipeline.py (hunk 2 above)
Comment (tombrooks248) [reply src/jobs/pipeline.py]: ok, found the duplices and made the code change to stop them appearing and also added an integration test which will hopefully catch any code changes which might lead to a similar issue in future.

File: src/jobs/pipeline.py (hunk 3 above)
Comment (tombrooks248) [reply src/jobs/pipeline.py]: Yep done.

File: src/jobs/pipeline.py (hunk 4 above)
Comment (tombrooks248) [reply src/jobs/pipeline.py:721]: Yep - refactored this so it looks like the others and most of the code goes in the Postgres writer utils file.

File: src/jobs/pipeline.py (hunk 5 above)
Comment (tombrooks248) [reply src/jobs/pipeline.py]: Good idea, yea made it do that.

File: src/jobs/pipeline.py (hunk 7 above)
Comment (tombrooks248) [reply src/jobs/pipeline.py]: Good idea, makes the code look so much cleaner.

**Review by eveleighoj — APPROVED** [review]
